### ISSUE TITLE
Add an optional FP8 KV-cache mode to Lily (Decode Only) 

### ISSUE THREAD

**Issue by rohash123** [perplexityai/pplx-garden#21]
## Problem

Lily requires an MLX affine 4-bit checkpoint, but its full-attention KV cache
is always stored in BF16. The cache dtype is hard-coded when each session state
is created:

https://github.com/perplexityai/pplx-garden/blob/main/lily/src/model.rs#L520-L560

There is currently no CLI option for choosing a KV-cache storage format. The
CLI exposes only `--model`, `--bind`, and `--max-seq`:

https://github.com/perplexityai/pplx-garden/blob/main/lily/src/main.rs

For Qwen3.6-35B-A3B, the ten full-attention layers use two KV heads with head
dimension 256. BF16 KV storage therefore costs approximately:

| `max_seq` | BF16 KV/session | Approx. 8-bit KV/session* |
|---:|---:|---:|
| 4,096 | 80 MiB | 40 MiB |
| 32,768 | 640 MiB | 320 MiB |
| 65,536 | 1.25 GiB | 640 MiB |
| 131,072 | 2.5 GiB | 1.25 GiB |
| 262,144 | 5 GiB | 2.5 GiB |

\*Before scale/alignment metadata.

## Proposed interface

Add an explicit cache-dtype option, initially preserving BF16 as the default:

```console
lily --kv-cache-dtype bf16
lily --kv-cache-dtype fp8
```

If the implementation uses scaled signed integers rather than a real FP8
encoding, the exposed value should be named `int8`, not `fp8`. If multiple FP8
encodings are supported, a more precise value such as `fp8-e4m3` would avoid
ambiguity.

## Tangible benefit

Mac CPUs, GPUs, Lily, and other applications share unified memory. An 8-bit
cache would approximately halve KV storage, allowing longer contexts with less
memory pressure. It could also reduce long-context decode memory traffic: each
decode step reads the existing K/V history, so halving bytes per cached value
may improve throughput when KV reads are bandwidth-bound.

Lily's published MLX 0.32.2 comparison shows decode throughput decreasing from
194.7 tok/s at context 256 to 92.7 tok/s at context 131,072:

https://github.com/perplexityai/pplx-garden/blob/main/lily/docs/2026-09-02-performance-mlx-0.32.2.md

That does not by itself prove KV bandwidth is the bottleneck, so throughput
improvement should be measured rather than assumed. The predictable benefit is
the roughly 50% reduction in KV payload bytes.

This is complementary to #20: dynamic or paged allocation avoids reserving
space for unused token positions, while an 8-bit format reduces the memory and
traffic for each allocated position.

## Implementation considerations

- Quantize K after Q/K normalization and RoPE, and quantize V before storing it.
- Dequantize inside the attention kernels; avoid materializing a complete BF16
  copy of the cache.
- Fail clearly when the requested representation is unsupported instead of
  silently falling back to BF16.
- Keep BF16 available as the accuracy/reference mode.
- **Keep prefill K/V in BF16 for now; apply INT8 KV storage only to decode, with the KV representation parameterized consistently across all decode attention variants.**

## Suggested validation

- Confirm the expected reduction in `MTLDevice.currentAllocatedSize` and process
  `phys_footprint`.
- Compare per-position logits against BF16 using fixed teacher-forced sequences
  at short and long context lengths.
- Measure maximum and aggregate logit error, and verify that it stays within an
  agreed tolerance.
- Benchmark decode throughput, especially at 32K–131K context, and verify that
  the BF16 reference path does not regress.
- Confirm that prefill remains on the existing BF16 path and has no performance
  or correctness regression.

**Comment by rohash123** [comment]
Happy to take this up if the team agrees!

**Comment by WetABQ** [comment]
Thanks, numbers check out. The benefit is real: at 131K each decode step streams
~2.5 GiB of K/V, and the 256 vs 131K step-time delta implies the attention path
is already close to memory bandwidth. Prefill will not benefit.

Before assigning, I would like a short note on prefill. The NAX prefill kernel
reads K/V via Metal 4 `tensor` views over device BF16 memory. There is no 8-bit
tensor dtype, so tiles must be dequantized into threadgroup memory, which is
capped at 32 KB while the current BK=128 K tile is 64 KB. Prefill is chunked, so
it cannot stay on BF16. Decode is tractable but the KV type should be a template
parameter across the split variants.

Two suggestions: go straight to `int8` with a per-(head, token) scale rather
than FP8, since MSL has no native fp8. And validate by comparing logits on a
teacher-forced sequence, not generation goldens; greedy trajectories diverge.

**Comment by rohash123** [comment]
Yeah, you’re right about prefill. It’s probably out of scope for this issue and best left in BF16 for now. We can narrow the proposal to an optional int8 KV cache for decode, using one scale per (head, token), rather than FP8, and template the KV type across the decode variants. For validation, we’ll compare logits on a teacher-forced sequence and skip generation goldens. I'll take this up!
